### PR TITLE
options: add --secondary-sub-pos

### LINKED ISSUE #3022: support for secondary subtitle streams not forced to top | Make secondary subtitle track style editable
MPV is one of the rare video players on Linux that supports reading several subtitle streams natively. I have some videos with several non-overlapping ASS subtitle tracks, and I would like to be able to display two of these tracks at the same time.

However, --secondary-sid is currently hardcoded to put the second track at the top of the video. Manually commenting out [this one line](https://github.com/mpv-player/mpv/blob/master/player/sub.c#L185) is enough to get rid of that behavior.

Would it be possible to have this be configurable? I'd like to avoid keeping local patches for my software. If you think it's worth it, I can volunteer my time to implement it.

### PR DESCRIPTION
**Pull request by karelrooted**
The default value is 0 (on the top of the screen)

related issue: #3022

### CODE REVIEW

File: DOCS/man/options.rst [hunk 1]
```diff
@@ -2338,8 +2338,8 @@ Subtitles
 ``--secondary-sid=<ID|auto|no>``
     Select a secondary subtitle stream. This is similar to ``--sid``. If a
     secondary subtitle is selected, it will be rendered as toptitle (i.e. on
-    the top of the screen) alongside the normal subtitle, and provides a way
-    to render two subtitles at once.
+    the top of the screen) alongside the normal subtitle by default, and
```
Comment (guidocella) [inline DOCS/man/options.rst:2341]: Trailing space

**Comment by guidocella** [comment]
The option works. The references to `SD_CTRL_SET_TOP` can be removed since it's no longer used.

File: player/command.c [hunk 2]
```diff
@@ -4249,6 +4262,7 @@ static const struct property_osd_display {
     {"sub", "Subtitles"},
     {"secondary-sid", "Secondary subtitles"},
     {"sub-pos", "Sub position"},
+    {"secondary-sub-pos", "Secondary Sub position"},
```
Comment (kasper93) [inline player/command.c:4265]:
```suggestion
    {"secondary-sub-pos", "Secondary sub position"},
```
to be consistent with `"Secondary sub delay"`

**Comment by karelrooted** [comment]
replace newline with space when display two subtitle at the bottom will have a better user experience, is adding another option [--sub-newline-to-space](https://github.com/karelrooted/mpv/commit/fe9e88d298af1e83caff5c6f9f53aae9fc78a03c) acceptable ?

**Comment by karelrooted** [comment]
> The option works. The references to `SD_CTRL_SET_TOP` can be removed since it's no longer used.

the unused SD_CTRL_SET_TOP is removed

**Comment by dyphire** [comment]
This has not completely closed #3022 , it still forces the subtitle style to be stripped  for secondary subtitles.
I tried to add a new option of [--secondary-sub-override](https://github.com/dyphire/mpv/commit/280e313e1f8ca3dde303bb0f8046f4253b18edc9) to control the subtitle style override for secondary subtitles, and submitted it after this PR was merged.

**Comment by karelrooted** [comment]
> This has not completely closed #3022 , it still forces the subtitle style to be stripped for secondary subtitles. I tried to add a new option of [--secondary-sub-override](https://github.com/dyphire/mpv/commit/cd6d1f8ed412af41571276338ef69ae53d413a36) to control the subtitle style override for secondary subtitles, and will submit it after this PR is merge.

 commit message is updated to no close #3022

**Comment by dyphire** [comment]
> > This has not completely closed #3022 , it still forces the subtitle style to be stripped for secondary subtitles. I tried to add a new option of [--secondary-sub-override](https://github.com/dyphire/mpv/commit/cd6d1f8ed412af41571276338ef69ae53d413a36) to control the subtitle style override for secondary subtitles, and will submit it after this PR is merge.
> 
> PR is updated to not close #3022 I mainly use this to play bilingual subtitle, only tested it on converted SRT subtitle, I think there is a reason secondary subtitle's ass style is being striped by default in the original design, so that behavior is maintained ( maybe when there is two ass subtitle ,usually there position style will cause them to overlap? )

The subtitle style that override secondary subtitles is meaningful, so the new option maintains its default behavior.

In certain scenarios, it can also be useful to preserve the style of ass subtitles for secondary subtitles. For example, some people prefer to preserve the style when combining bilingual ass subtitles, while others may obtain Danmaku within ass subtitle format when playing videos from websites such as niconico and Bilibili. They also need to preserve the style to obtain the correct rendering when displaying as secondary subtitles.

**Comment by dyphire** [comment]
> > This has not completely closed #3022 , it still forces the subtitle style to be stripped for secondary subtitles. I tried to add a new option of [--secondary-sub-override](https://github.com/dyphire/mpv/commit/cd6d1f8ed412af41571276338ef69ae53d413a36) to control the subtitle style override for secondary subtitles, and will submit it after this PR is merge.
> 
> PR is updated to not close #3022 I mainly use this to play bilingual subtitle, only tested it on converted SRT subtitle, I think there is a reason secondary subtitle's ass style is being striped by default in the original design, so that behavior is maintained ( maybe when there is two ass subtitle ,usually there position style will cause them to overlap? )

The subtitle style that override secondary subtitles is meaningful, so the new option maintains its default behavior.

In certain scenarios, it can also be useful to preserve the style of ass subtitles for secondary subtitles. For example, some people prefer to preserve the style when combining bilingual ass subtitles, while others may obtain Danmaku within ass subtitle format when playing videos from websites such as niconico and Bilibili. They also need to preserve the style to obtain the correct rendering when displaying as secondary subtitles.

**Comment by Lypheo** [comment]
Btw, it’s actually [quite trivial](https://github.com/Lypheo/mpv/commit/51ac3b3999c4d11a66f4c893b8b9a927c494d05f) to make secondary subs fully configurable, but I’m not sure how to do it without all that ugly code duplication.

**Comment by Dudemanguy** [comment]
You can just do this to duplicate it. Of course you would need to adjust `sec_sub_visibility` and `sub_delay[2]` as well. The only weird thing is that this actually hits an assert that was introduced in 4cb264a3ff35cd1dcd8a41f87fb87791920d5156. I'm not really sure if that's needed.

```
diff --git a/options/m_config_core.c b/options/m_config_core.c
index 08a76eb76d..27534b817c 100644
--- a/options/m_config_core.c
+++ b/options/m_config_core.c
@@ -466,8 +466,8 @@ static void add_sub_group(struct m_config_shadow *shadow, const char *name_prefi
                           const struct m_sub_options *subopts)
 {
     // Can't be used multiple times.
-    for (int n = 0; n < shadow->num_groups; n++)
-        assert(shadow->groups[n].group != subopts);
+    //for (int n = 0; n < shadow->num_groups; n++)
+    //    assert(shadow->groups[n].group != subopts);
 
     if (!name_prefix)
         name_prefix = "";
diff --git a/options/options.c b/options/options.c
index 41cfce7fc0..ce8096e186 100644
--- a/options/options.c
+++ b/options/options.c
@@ -284,11 +284,9 @@ const struct m_sub_options mp_sub_filter_opts = {
 const struct m_sub_options mp_subtitle_sub_opts = {
     .opts = (const struct m_option[]){
         {"sub-delay", OPT_FLOAT(sub_delay[0])},
-        {"secondary-sub-delay", OPT_FLOAT(sub_delay[1])},
         {"sub-fps", OPT_FLOAT(sub_fps)},
         {"sub-speed", OPT_FLOAT(sub_speed)},
         {"sub-visibility", OPT_BOOL(sub_visibility)},
-        {"secondary-sub-visibility", OPT_BOOL(sec_sub_visibility)},
         {"sub-forced-events-only", OPT_BOOL(sub_forced_events_only)},
         {"stretch-dvd-subs", OPT_BOOL(stretch_dvd_subs)},
         {"stretch-image-subs-to-screen", OPT_BOOL(stretch_image_subs)},
@@ -667,6 +665,7 @@ static const m_option_t mp_opts[] = {
     {"cover-art-whitelist", OPT_BOOL(coverart_whitelist)},
 
     {"", OPT_SUBSTRUCT(subs_rend, mp_subtitle_sub_opts)},
+    {"secondary", OPT_SUBSTRUCT(subs2_rend, mp_subtitle_sub_opts)},
     {"", OPT_SUBSTRUCT(subs_filt, mp_sub_filter_opts)},
     {"", OPT_SUBSTRUCT(osd_rend, mp_osd_render_sub_opts)},
 
diff --git a/options/options.h b/options/options.h
index f1940266ca..d2c66d4d37 100644
--- a/options/options.h
+++ b/options/options.h
@@ -198,6 +198,7 @@ typedef struct MPOpts {
     bool cursor_autohide_fs;
 
     struct mp_subtitle_opts *subs_rend;
+    struct mp_subtitle_opts *subs2_rend;
     struct mp_sub_filter_opts *subs_filt;
     struct mp_osd_render_opts *osd_rend;
```

Of course adjust the

**Comment by karelrooted** [comment]
> Btw, it’s actually [quite trivial](https://github.com/Lypheo/mpv/commit/51ac3b3999c4d11a66f4c893b8b9a927c494d05f) to make secondary subs fully configurable, but I’m not sure how to do it without all that ugly code duplication.

fully configurable sounds good, I will close this PR in favor of your version

**Comment by dyphire** [comment]
Is it really necessary to implement all `sub-*` options for secondary subtitles? I am skeptical about the practicality of my comparison

**Comment by Dudemanguy** [comment]
Yeah I'm not really sure all of them are needed either. We do keep getting people wanting to customize things for secondary subtitles however so it probably makes sense to generalize somewhat instead of constantly adding yet another --secondary option.

**Comment by dyphire** [comment]
> Yeah I'm not really sure all of them are needed either. We do keep getting people wanting to customize things for secondary subtitles however so it probably makes sense to generalize somewhat instead of constantly adding yet another `--secondary-*` option.

This also sounds reasonable. If we want to add fully configurable `--secondary-*` options for secondary subtitles, please also note that setting their default values separately for `--secondary-sub-visibility` and `--secondary-sub-ass-override` options to maintain the current default behavior. The above example changes missed this point.

**Comment by Lypheo** [comment]
> fully configurable sounds good, I will close this PR in favor of @Lypheo 's fully configurable version

Don’t, I didn’t intend to make a PR. You can incorporate it into yours if you want.

**Comment by Dudemanguy** [comment]
Given that secondary-sub-pos is a bit special since it also removes some sd_ctrls. I think we can get this sorted out first and then refactor it later.

**Review by Dudemanguy — APPROVED** [review]
Works fine

**Comment by Dudemanguy** [comment]
I will follow up with a small refactor shortly.

**Pull request merged**